main.py

```diff
@@ -35,91 +43,200 @@ def save_file(path, pred, startrows=None, endrows=None):
     # startRow – first row to save
     # endRow – last row to load, or the last of file, the minimum
     # TODO:find better way to partially save file,
-    #  dont want to take too much memory, or time
+    #  dont want to take too much memory
+    temp = load_data(path, startrows, endrows)
+    temp["grouping"] = pred.tolist()
     if path[-5:] == ".xlsx":
-        temp = pd.read_excel(path, nrows=endrows - startrows, skiprows=startrows)
         new_path = path[:-5] + "_new" + path[-5:]
-        temp = pd.concat(temp, pred)
-        temp.to_excel(new_path)
+        temp.to_excel(new_path,mode='a')# temp.to_excel(new_path,index=[i for i in range(startrows,endrows)])
     elif path[-4:] == ".xls":
-        temp = pd.read_excel(path, nrows=endrows - startrows, skiprows=startrows)
         new_path = path[:-4] + "_new" + path[-4:]
-        temp = pd.concat(temp, pred)
-        temp.to_excel(new_path)
+        temp.to_excel(new_path,mode='a')
     elif path[-4:] == ".csv":
-        temp = pd.read_csv(path, nrows=endrows - startrows, skiprows=startrows)
         new_path = path[:-4] + "_new" + path[-4:]
-        temp = pd.concat(temp, pred)
-        temp.to_csv(new_path)
+        temp.to_csv(new_path,mode='a')
 
 
-#__________________________________________
-#from here on is a bit of a mess, dont read if you value your sanity
-#and by mess i mean incomplete functions and trial and error
-#__________________________________________
+def decide_row_num_halt(path,cols=None):
+    #decide how many rows can you work with at atime, depends on the DATABUFFER
+    if path[-5:] == ".xlsx" or path[-4:] == ".xls":
+        temp = pd.read_excel(path,usecols=cols,nrows=1)
+    elif path[-4:] == ".csv":
+        temp=pd.read_csv(path,usecols=cols,nrows=1)
+    return round(DATABUFFER/ sys.getsizeof(temp))
 
-def get_time_replace_reduct_cols(data):
+
+def get_time_replace_reduct_cols(path,rowNum):
     #get_time_replace_reduct_cols(data) -> timeCols,replaceDict,colNums
     # data – data we work with currently
     # timeCols – all columns with timestamps
     # replaceDict  - list of columns and dictionaries to replace their values with
     # colNums – list of columns to load, as some of them are useless, example:  personal numbers, passwords and OTP
-    timeCols= []
-    replaceDict = []
-    colNums = []
-    # row = data[:][0]
-    # colLen = len(data)
-    # for i,val in enumerate(row):
-    #     #check for OTP
-    #     if type(val) == 'string':
-    #         things =[]
-    #         for j in data[:][i]:
-    #             if not (j in things):
-    #                 if len(things)>= colLen/2:
-    #                     break
-    #                 things.insert(0,j)
-
 
+    data = load_data(path, endrows=rowNum)
+    timeCols = []
+    replaceDict = {}
+    colNums = []
+    numberTypes = numpy.array(["int8","int16","int32","int64","uint8","uint16","uint64","uint32","float32","float64"])
```

Comment:
Surely there must be a nicer way. 

But if there isn't (there is) - at least make it a constant